libbeat/processors/add_host_metadata/add_host_metadata.go

```diff
@@ -134,38 +134,40 @@ func (p *addHostMetadata) Run(event *beat.Event) (*beat.Event, error) {
 //}
 
 func (p *addHostMetadata) expired() bool {
-	if p.config.CacheTTL <= 0 {
-		return true
-	}
 
 	p.lastUpdate.Lock()
 	defer p.lastUpdate.Unlock()
 
+	if p.config.CacheTTL <= 0 {
+		return true
+	}
+
 	if p.lastUpdate.Add(p.config.CacheTTL).After(time.Now()) {
 		return false
 	}
 	p.lastUpdate.Time = time.Now()
```

Comment:
It looks to me like this is still a race -- if the cache is expired, then calling `expired()` twice in a row will return true the first time and false the second time. But to give a correct result, once `expired()` becomes true, it should stay true until the next `loadData` completes.

The PR description mentions not wanting everything to grind to a halt on every `loadData`, and that makes sense, but everything _should_ grind to a halt when the cache is expired, until at least one `loadData` completes. The tricky part here is making sure that we don't do redundant work -- the first caller to notice that the cache is expired should call `loadData`, then any additional callers should notice that the cache is expired and the result is already being recomputed, and wait for it to finish.

Reply:
So, my thinking is that with `loadData()`,  there's a trade off involved:

Normal cache expiry doesn't hold the mutex: A certain number of events get stale hostname data, but updating the hostname will only block one event.
Normal Cache expiry holds the mutex: No events get stale hostname data, but we globally block events until the hostname is updated.

The latter case makes me pretty nervous. Note that the call to `FQDN()` that happens in `loadData()` can, depending on the platform & setup, involve some network calls to look up DNS info. Blocking all event ingestion while we do network calls is kinda scary, so I figured the tradeoff was worth it. Open to other ideas, though.

Reply:
Hmm, so one simplifying factor that I think we weren't relying on: a processor is run on a single goroutine. On the one hand, this means that some things that look like race conditions aren't: updating the cache TTL in `expired` when it is checked rather than in `loadData` when it is updated makes the code confusing, but it doesn't actually break anything, because the only important caller of `expired()` is `loadData`, and if `expired()` returns true then `loadData` synchronously reloads everything.

(Important exception: the tests. The existing tests used `expired()` to "prime" the cache's lastUpdated value, even though `expired()` itself doesn't load any data. If the tests need a special call to mark the cache as non-expired, then there should be separate helper functions for checking expiration and resetting expiration time, instead of making `expired()` itself do both, otherwise at the very least it's hard to reason about whether they're testing the right thing.)

The upshot though is: the internal behavior was confusing but sequentially correct. The error comes in with FQDN flag updates, which are the only changes that happen on a different goroutine than the processor itself. The previous behavior was _almost_ correct: while holding a mutex, set the cache update time to zero, so the next call to `expired()` would return true, which would trigger a synchronous update of FQDN on the next `loadData()` call. Since the processor itself is locked to a single goroutine, no other callers will try using `loadData()` while that update is happening.

I think the core problem might be the callback issue I pointed out elsewhere: the FQDN callbacks overwrite each other, so only one copy of `add_host_metadata` receives the update, and the rest just have to wait until their `cacheTTL` expires.

Reply:
So, you're right that the behavior is "almost" correct. When I was still trying to figure out a fix, one thing I experimented with is how many data streams it takes to reproduce the issue. If you just have one datastream installed, you don't hit the bug, as there's only one goroutine involved, like you said. Issue happens when we have multiple data streams, and thus multiple threads.

Not sure what you mean by "the FQDN callbacks overwrite each other". Do you mean how the cache expiry happens between threads?

Reply:
Ah, okay. I did change that. Now it returns an error, and the processor will use a keyname with the UUID.

I could go back and run some tests, but my understanding is that that's largely unrelated, and that under agentonly one processor instance will register itself [during startup](https://github.com/elastic/beats/blob/7a59da36198357e492887af44dc465d9bee24a69/x-pack/metricbeat/cmd/root.go#L68)